src/target/riscv/riscv-013.c

```diff
@@ -3146,10 +3130,33 @@ static int modify_privilege(struct target *target, uint64_t *mstatus, uint64_t *
 	*mstatus = set_field(*mstatus, MSTATUS_MPRV, 1);
 
 	/* Write MSTATUS */
-	if (*mstatus == *mstatus_old)
+	if (*mstatus != *mstatus_old &&
+			register_write_direct(target, GDB_REGNO_MSTATUS, *mstatus) != ERROR_OK)
```

Comment:
Consider also using `register_read/write_direct()` in `modify_privilege()`.
The issue is: what if the write of `mstatus` succeeds but the write of `dcsr` fails? This will change the value of `mstatus` on the target without invalidating the cached value.

Reply:
address